runtime/runtime-standalone/src/lib.rs

```diff
@@ -0,0 +1,394 @@
+    pub gas_price: Balance,
+    pub gas_limit: Gas,
+}
+
+impl Block {
+    pub fn genesis(genesis_config: &GenesisConfig) -> Self {
+        Self {
+            prev_block: None,
+            state_root: CryptoHash::default(),
+            block_height: 1,
+            epoch_height: 1,
+            block_timestamp: genesis_config.genesis_time,
+            gas_price: genesis_config.gas_price,
+            gas_limit: genesis_config.gas_limit,
+            gas_burnt: 0,
+        }
+    }
+
+    pub fn produce(&self, new_state_root: CryptoHash) -> Block {
+        Self {
+            gas_price: self.gas_price,
+            gas_limit: self.gas_limit,
+            block_timestamp: self.block_timestamp + 1,
+            prev_block: Some(Box::new(self.clone())),
+            state_root: new_state_root,
+            block_height: self.block_height + 1,
+            epoch_height: self.epoch_height + 1, // TODO: simulate epoch_height
+            gas_burnt: 0,
+        }
+    }
+}
+
+pub struct RuntimeStandalone {
+    tx_pool: TransactionPool,
+    transactions: HashMap<CryptoHash, SignedTransaction>,
+    outcomes: HashMap<CryptoHash, ExecutionOutcome>,
+    cur_block: Block,
+    runtime: Runtime,
+    trie: Arc<Trie>,
+    pending_receipts: Vec<Receipt>,
+}
+
+impl RuntimeStandalone {
+    pub fn new(genesis: GenesisConfig, store: Arc<Store>) -> Self {
+        let mut genesis_block = Block::genesis(&genesis);
+        let mut store_update = store.store_update();
+        let runtime = Runtime::new(genesis.runtime_config.clone());
+        let trie = Arc::new(Trie::new(store));
+        let trie_update = TrieUpdate::new(trie.clone(), CryptoHash::default());
+        let (s_update, state_root) =
+            runtime.apply_genesis_state(trie_update, &[], &genesis.state_records);
+        store_update.merge(s_update);
+        store_update.commit().unwrap();
+        genesis_block.state_root = state_root;
+        Self {
+            trie,
+            runtime,
+            transactions: HashMap::new(),
+            outcomes: HashMap::new(),
+            cur_block: genesis_block,
+            tx_pool: TransactionPool::new(),
+            pending_receipts: vec![],
+        }
+    }
+
+    /// Processes blocks until the final value is produced
+    pub fn resolve_tx(
+        &mut self,
+        mut tx: SignedTransaction,
+    ) -> Result<ExecutionOutcome, RuntimeError> {
+        tx.init();
+        let mut outcome_hash = tx.get_hash();
+        self.transactions.insert(outcome_hash, tx.clone());
+        self.tx_pool.insert_transaction(tx);
+        loop {
+            self.process_block()?;
+            let outcome = self.outcomes.get(&outcome_hash).unwrap();
+            match outcome.status {
+                ExecutionStatus::SuccessReceiptId(ref id) => outcome_hash = *id,
+                ExecutionStatus::SuccessValue(_)
+                | ExecutionStatus::Failure(_)
+                | ExecutionStatus::Unknown => return Ok(outcome.clone()),
```

Comment:
If the status is `Unknown` it means that transaction hasn't been processed so I think we should not return here. Also the way you write it, the outcome will be the first receipt outcome, which I am not sure is intended.

Reply:
> Also the way you write it, the outcome will be the first receipt outcome, which I am not sure is intended.

I'm not sure if I'm understand correctly what do you mean. In this code we are waiting for the final outcome (which is not SuccessReceiptId).

Reply:
It depends on what you mean by "final outcome". Two things:
* `Unknown` cannot be the final outcome. We use `Unknown` to indicate that the processing hasn't started yet.
* You would return the receipt outcome here and the transaction outcome will be discarded. Depending on what you want, this may be fine.

Reply:
> We use `Unknown` to indicate that the processing hasn't started yet.

Really unfortunate naming...

Reply:
As I read it, `Unknown` means that we need to continue processing another block:

```rust
        loop {
            self.process_block()?;
            let outcome = self.outcomes.get(&outcome_hash).unwrap();
            match outcome.status {
                ExecutionStatus::Unknown => {},
                ExecutionStatus::SuccessReceiptId(ref id) => outcome_hash = *id,
                ExecutionStatus::SuccessValue(_)
                | ExecutionStatus::Failure(_) => return Ok(outcome.clone()),
            };
        }
```

@bowenwang1996 Is that correct?

Reply:
yes

Reply:
@fckt see what @frol suggested

Reply:
@bowenwang1996 @fckt wrote below:

> > As I read it, Unknown means that we need to continue processing another block
>
> @frol runtime is fully deterministic, it never outputs outcomes with "Unknown" status.

:thinking: 

If that is true, we should address the proper type-encoding, and mark this `Unknown` status as `unreachable!` instead of silently propagating impossible state up.

Reply:
good point

Reply:
addressed, thanks